### PR TITLE
fix: close query on invalid use of HTTP/2 with /query endpoint

### PR DESCRIPTION
**Pull request by vcrfxia**
### Description 

Fixes https://github.com/confluentinc/ksql/issues/5881.

The bug is that the close callback passed to the QueryStreamWriter/TopicStreamWriter sets a flag (`connectionClosed`) that causes the `write()` loop to end and close the underlying Kafka Streams app/Kafka consumer, but when HTTP/2 is used, `write()` is never called so this doesn't happen even though the `connectionClosed` flag is set. This PR fixes the bug by introducing a new method, `closeWithoutWrite()`, that directly closes the underlying Kafka Streams app/Kafka consumer and is called when the 405 response is returned (upon discovering an HTTP/2 request to the /query endpoint for a request type that requires an HTTP streaming response).

### Testing done 

Manual + integration tests.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by apurvam** [comment]
@purplefox can you help review this?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/OldApiUtils.java [hunk 1]
```diff
@@ -89,15 +89,16 @@ static void handleOldApiResponse(final Server server, final RoutingContext routi
     // a plain String, other times it's an object that needs to be JSON encoded, other times
     // it represents a stream.
     if (endpointResponse.getEntity() instanceof StreamingOutput) {
+      final StreamingOutput streamingOutput = (StreamingOutput) endpointResponse.getEntity();
       if (routingContext.request().version() == HttpVersion.HTTP_2) {
         // The old /query endpoint uses chunked encoding which is not supported in HTTP2
         routingContext.response().setStatusCode(METHOD_NOT_ALLOWED.code())
             .setStatusMessage("The /query endpoint is not available using HTTP2").end();
+        streamingOutput.closeWithoutWrite();
```
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/OldApiUtils.java:97]: did you mean to change this to `close`?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/StreamingOutput.java [hunk 2]
```diff
@@ -21,4 +21,7 @@
 public interface StreamingOutput {
 
   void write(OutputStream output) throws IOException;
+
+  void close();
```
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/StreamingOutput.java:25]: nit, we should instead have this `implement AutoCloseable`

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStreamWriter.java [hunk 3]
```diff
@@ -137,7 +138,15 @@ public void write(final OutputStream out) {
       log.error("Exception encountered while writing to output stream", exception);
       outputException(out, exception);
     } finally {
+      close();
+    }
+  }
+
+  @Override
+  public void close() {
```
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStreamWriter.java:146]: synchronized?

Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStreamWriter.java:146]: Argh, yes. Thanks for the catch!

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/OldApiUtils.java (hunk 1 above)
Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/OldApiUtils.java:97]: 😅

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/StreamingOutput.java (hunk 2 above)
Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/StreamingOutput.java:25]: Are you suggesting the classes implement AutoCloseable, or that the interface extends AutoCloseable? The latter makes sense to but though I got a compiler warning:
```
Warning:(21, 8) java: auto-closeable resource io.confluent.ksql.api.server.StreamingOutput has a member method close() that could throw InterruptedException
```
I worked around it with the same workaround we use in ServiceContext.java. LMK if this isn't what you had in mind.

Comment (agavra) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/StreamingOutput.java:26]: eh I was suggesting the latter but I'm not entirely sure I understand the warning.

Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/StreamingOutput.java:26]: I don't fully understand it either but it seems general consensus is that the warning is bogus and the workaround we've used is fine:
https://bugs.openjdk.java.net/browse/JDK-8155591
https://bugs.openjdk.java.net/browse/JDK-7027157
https://issues.apache.org/jira/browse/CXF-7465
https://github.com/joyent/java-manta/issues/322

Going to leave this as is.

**Pull request merged**